test/PolymorphicFields.purs

```diff
@@ -56,3 +70,88 @@ test = do
     openConsumer { x: true, y: false, z: 42 } == show [true, false]
   assert $ -- Make sure explicitly wrapping the field in `Req` also works
     openConsumer { x: Req 5, z: 42 } == show [5]
+  assert $
+    nestedOpenConsumer 42 == show [42, 42]
+
+  assert $
+    monoConsumer { x: 42, z: 42 } == show [42]
+  assert $
+    monoConsumer { x: 42, y: 5, z: 42 } == show [42, 5]
+  assert $
+    monoAmbiguousConsumer { x: Just 42, z: 42 } == show [Just 42]
+  assert $
+    monoAmbiguousConsumer { x: Just 42, y: Just 5, z: 42 } == show [Just 42, Just 5]
+  assert $
+    monoAmbiguousConsumer { x: Nothing, y: Just 5, z: 42 } == show [Nothing, Just 5]
+  assert $
+    monoAmbiguousConsumer { x: Just 42, y: Nothing, z: 42 } == show [Just 42, Nothing]
+
+-- This test explicitly enumerates all coercion use cases that we want to
+-- support. If instance resolution breaks, one or more lines in this function
+-- will fail to compile.
+compileTimeTestClosed :: Identity Unit
+compileTimeTestClosed = do
+  witness (Proxy :: _ Int) (Proxy :: _ Int)
+  witness (Proxy :: _ (Req Int)) (Proxy :: _ Int)
+  witness (Proxy :: _ (Opt Int)) (Proxy :: _ Int)
+  witness (Proxy :: _ (Opt Int)) (Proxy :: _ (Opt Int))
+  witness (Proxy :: _ (Opt (Maybe Int))) (Proxy :: _ (Maybe Int))
+  witness (Proxy :: _ (Opt (Maybe Int))) (Proxy :: _ (Opt (Maybe Int)))
+  genericWitness (Proxy :: _ Int)
+  witness (Proxy :: _ { x :: Req Int, y :: Opt Int }) (Proxy :: _ { x :: Int, y :: Int })
```

Comment:
Also added a set of tests explicitly enumerating the coercion use cases that should be supported.